### PR TITLE
feat: uploading to prefix.dev

### PR DESCRIPTION
**Pull request by wolfv**

### CODE REVIEW

File: src/upload.rs [hunk 1]
```diff
@@ -70,3 +94,107 @@ pub async fn upload_package_to_quetz(
...
+        .body(body)
+        .send()
+        .await
+        .into_diagnostic()?;
+
+    if response.status().is_success() {
+        println!("Upload successful!");
+    } else {
+        println!("Upload failed!");
+        if response.status() == 401 {
```
Comment (0xbe7a) [inline src/upload.rs:189]: You can use https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.error_for_status

**Comment by wolfv** [comment]
@0xbe7a I refactored the code a little bit and added progress bars to each of the methods. Do you think you can test this for quetz & artifactory?

This also allows to put more than a single package for uploading, e.g.

`rattler-build upload prefix --channel my-test *.conda`

File: src/upload.rs [hunk 2]
```diff
@@ -115,37 +131,150 @@ pub async fn upload_package_to_artifactory(
...
+            .into_diagnostic()
+            .wrap_err("Creating temporary directory failed")?;
+
+        rattler_package_streaming::fs::extract(&package_file, package_dir.path())
+            .into_diagnostic()?;
+
+        let index_json = IndexJson::from_package_directory(package_dir.path()).into_diagnostic()?;
+        let subdir = index_json
+            .subdir
+            .ok_or_else(|| miette::miette!("index.json of the package has no subdirectory. Cannot determine which directory to upload to"))?;
```
Comment (0xbe7a) [inline src/upload.rs:145]: We should probably include the package name in the error message

**Comment by 0xbe7a** [comment]
Looks good, i will test the Quetz and Artifactory upload later today

**Comment by wolfv** [comment]
OK, let me know how it goes and then I'll merge :)

**Comment by 0xbe7a** [comment]
Works as advertised, ready to merge! 🤠

**Pull request merged**